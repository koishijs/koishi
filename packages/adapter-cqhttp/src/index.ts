import { Server, Session, App } from 'koishi-core'
import { Logger } from 'koishi-utils'
import HttpServer from './http'
import WsClient from './ws'
import WsServer from './ws-reverse'

export * from './api'
export * from './channel'
export * from './http'
export * from './ws'
export * from './ws-reverse'

declare module 'koishi-core/dist/server' {
  interface BotOptions {
    token?: string
  }
}

Server.types['cqhttp:http'] = HttpServer
Server.types['cqhttp:ws'] = WsClient
Server.types['cqhttp:ws-reverse'] = WsServer

const logger = Logger.create('app')

interface CQHTTP extends Server {}

class CQHTTP {
  constructor (app: App) {
    const bot = app.options.bots.find(bot => bot.server)
    if (!bot) {
      logger.info('infer type as %c', 'cqhttp:ws-reverse')
      return new WsServer(app)
    }
    if (bot.server.startsWith('ws')) {
      logger.info('infer type as %c', 'cqhttp:ws')
      return new WsClient(app)
    } else {
      logger.info('infer type as %c', 'cqhttp:http')
      return new HttpServer(app)
    }
  }
}

// register default behavior
Server.types.cqhttp = CQHTTP
Server.types.undefined = CQHTTP

Session.prototype.$send = async function $send (this: Session, message: string, autoEscape = false) {
  if (!message) return
  let ctxId: number
  // eslint-disable-next-line no-cond-assign
  const ctxType = (ctxId = this.groupId) ? 'group' : (ctxId = this.userId) ? 'user' : null
  if (this.$app.options.preferSync) {
    ctxType === 'group'
      ? await this.$bot.sendGroupMsg(ctxId, message, autoEscape)
      : await this.$bot.sendPrivateMsg(ctxId, message, autoEscape)
    return
  }
  if (this._response) {
    const session = this.$bot.createSession(this.messageType, ctxType, ctxId, message)
    if (this.$app.bail(this, 'before-send', session)) return
    return this._response({ reply: message, autoEscape, atSender: false })
  }
  return ctxType === 'group'
    ? this.$bot.sendGroupMsgAsync(ctxId, message, autoEscape)
    : this.$bot.sendPrivateMsgAsync(ctxId, message, autoEscape)
}
