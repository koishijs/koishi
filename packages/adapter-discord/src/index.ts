import { Adapter } from 'koishi-core'
import { AxiosRequestConfig } from 'axios'
import { DiscordBot } from './bot'
import WsClient from './ws'
import * as dc from './types'
export * from './bot'
export * as dc from './types'

interface DiscordOptions extends Adapter.WsClientOptions {
  axiosConfig?: AxiosRequestConfig
}

declare module 'koishi-core' {
  interface AppOptions {
    discord?: DiscordOptions
  }

   interface Session {
    discord?: {
      mentions: dc.User[];
      // eslint-disable-next-line camelcase
      webhook_id?: dc.snowflake
      flags: number
    }
  }

  namespace Bot {
    interface Platforms {
      discord: DiscordBot
    }
  }
}

Adapter.types['discord'] = WsClient
