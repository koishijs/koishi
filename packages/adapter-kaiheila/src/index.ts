import { App, Server } from 'koishi-core'
import { Logger, defineProperty, assertProperty, camelize, renameProperty } from 'koishi-utils'
import { createSession, KaiheilaBot } from './bot'
import { createHmac } from 'crypto'

interface KaiheilaOptions {
  path?: string
}

declare module 'koishi-core/dist/app' {
  interface AppOptions {
    kaiheila?: KaiheilaOptions
  }
}

declare module 'koishi-core/dist/server' {
  interface BotOptions {
    verifyToken?: string
    token?: string
  }
}

type ChannelType = 'GROUP' | 'WEBHOOK_CHALLENGE'

enum EventType {
  text = 1,
  image = 2,
  video = 3,
  file = 4,
  audio = 8,
  markdown = 9,
  system = 255,
}

interface Payload {
  s: number
  sn?: number
  d: {
    type: EventType
    channelType: ChannelType
    challenge: string
    verifyToken: string
    targetId: string
    authorId: string
    content: string
    msgId: string
    msgTimestamp: number
    nonce: string
    extra: {
      type: string
      guildId: string
      channelName: string
      mention: []
      mentionAll: boolean
      mentionRoles: []
      mentionHere: boolean
      code: string
      author: {
        identifyNum: string
        avatar: string
        username: string
        id: string
        nickname: string
        roles: []
      }
      attachments: {
        type: string
        url: string
        name: string
        fileType: string
        size: number
        duration: number
        width: number
        height: number
      },
      body: {
        id: string
        username: string
        identifyNum: string
        online: boolean
        os: string
        status: number
        avatar: string
        mobileVerified: boolean
        nickname: string
        roles: []
        joinedAt: number
        activeTime: number
      }
    }
  }
}

const logger = new Logger('server')

Server.types.kaiheila = class HttpServer extends Server<KaiheilaBot> {
  constructor(app: App) {
    assertProperty(app.options, 'port')
    super(app, KaiheilaBot)
  }

  private async _listen(bot: KaiheilaBot) {
    bot.ready = true
  }

  async listen() {
    const { kaiheila = {} } = this.app.options
    const { path = '/' } = kaiheila
    this.app.router.post(path, (ctx) => {
      logger.debug('receive %o', ctx.request.body)
      const data = camelize<Payload>(ctx.request.body)

      const { challenge } = data.d
      ctx.body = { challenge }
      ctx.status = 200

      const bot = this.bots.find(bot => bot.verifyToken === data.d.verifyToken)
      if (!bot) return

      const meta = data.d as any
      KaiheilaBot.toMessage(meta)
      const session = createSession(this.app, {
        ...data.d,
        selfId: bot.selfId,
        userId: data.d.authorId,
        eventType: 'message',
        kind: 'kaiheila',
        channelId: data.d.targetId,
        subType: data.d.targetId ? 'group' : 'private',
      })

      // dispatch events
      console.log(session)
      this.dispatch(session)
    })

    await Promise.all(this.bots.map(bot => this._listen(bot)))
  }

  close() {
    logger.debug('http server closing')
  }
}
