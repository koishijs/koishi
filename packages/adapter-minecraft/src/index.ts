import { Adapter } from 'koishi-core'
import { MinecraftBot } from './bot'
import WsClient from './ws'
export * from './bot'

declare module 'koishi-core' {
  interface AppOptions {
    minecraft?: {}
  }

  namespace Bot {
    interface Platforms {
      minecraft: MinecraftBot
    }
  }
}

Adapter.types['minecraft'] = WsClient
