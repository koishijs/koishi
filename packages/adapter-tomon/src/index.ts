import SocketSession from './network/session'
import { App, EventType, Server, Session } from 'koishi-core'
import { TomonBot, TomonMessageInfo } from './bot'
import { camelize } from 'koishi-utils'

export * from './bot'

declare module 'koishi-core/dist/server' {
  interface BotOptions {
    token?: string
    fullName?: string
    password?: string
  }
}

Server.types.tomon = class TomonServer extends Server<TomonBot> {
  constructor(app: App) {
    super(app, TomonBot)
  }

  async _listen(bot: TomonBot) {
    // part 1: authorization
    const { token, fullName, password } = bot
    const info = await bot.request('POST', '/auth/login', {
      data: { token, fullName, password },
      auth: false,
    })
    Object.assign(bot, info)
    bot.ready = true
    const selfId = bot.selfId = bot.id

    // part 2: connect to server
    return new Promise<void>((resolve) => {
      const socket = new SocketSession()
      socket.token = bot.token
      socket.open()

      socket._emitter.once('READY', () => {
        console.log(`Bot ${bot.name}(${bot.username}#${bot.discriminator}) is ready to work!`)
        resolve()
      })

      const dispatchMessage = (data: TomonMessageInfo, eventType: EventType) => {
        TomonBot.toMessage(data = camelize(data))
        const userId = data.author.id
        if (userId === selfId) return
        // TODO: 处理图片和表情
        if (!data.content) data.content = ''
        this.dispatch(new Session(this.app, {
          ...data,
          selfId,
          userId,
          eventType,
          kind: 'tomon',
          channelId: data['guildId'],
          subType: data['guildId'] ? 'group' : 'private',
        }))
      }

      socket._emitter.on('MESSAGE_CREATE', async ({ d }) => dispatchMessage(d, 'message'))
      socket._emitter.on('MESSAGE_UPDATE', async ({ d }) => dispatchMessage(d, 'message-updated'))
      socket._emitter.on('MESSAGE_DELETE', async ({ d }) => dispatchMessage(d, 'message-deleted'))
    })
  }

  async listen() {
    await Promise.all(this.bots.map(bot => this._listen(bot)))
  }

  close() {}
}
