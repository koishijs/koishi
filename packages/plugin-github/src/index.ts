/* eslint-disable camelcase */

import { Context, User } from 'koishi-core'
import { Webhooks, EventNames } from '@octokit/webhooks'
import { GetWebhookPayloadTypeFromEvent } from '@octokit/webhooks/dist-types/generated/get-webhook-payload-type-from-event'
import axios from 'axios'

type Payload<T extends EventNames.All> = GetWebhookPayloadTypeFromEvent<T, unknown>['payload']

declare module 'koishi-core/dist/database' {
  interface User {
    githubToken?: string
  }
}

User.extend(() => ({
  githubToken: '',
}))

export interface Config {
  secret?: string
  webhook?: string
  authorize?: string
  clientId?: string
  clientSecret?: string
  redirectUri?: string
  repos?: Record<string, number[]>
}

const defaultOptions: Config = {
  secret: '',
  webhook: '/github/webhook',
  authorize: '/github/authorize',
  repos: {},
}

export const name = 'github-webhook'

export function apply(ctx: Context, config: Config = {}) {
  if (!ctx.router) throw new Error('ctx.router is not defined')

  config = { ...defaultOptions, ...config }
  const webhook = new Webhooks(config)

  const { router, database } = ctx

  router.post(config.webhook, (ctx, next) => {
    return webhook.middleware(ctx.req, ctx.res, next)
  })

  router.get(config.authorize, async (ctx) => {
    const targetId = parseInt(ctx.query.state)
    if (Number.isNaN(targetId)) throw new Error('Invalid targetId')
    const { code, state } = ctx.query
    const { data } = await axios.post('https://github.com/login/oauth/access_token', {
      client_id: config.clientId,
      client_secret: config.clientSecret,
      redirect_uri: config.redirectUri,
      code,
      state,
    })
    if (data.access_token) {
      await database.setUser(targetId, { githubToken: data.access_token })
      return ctx.status = 200
    } else {
      return ctx.status = 500
    }
  })

  function registerHandler<T extends EventNames.All>(event: T, handler: (payload: Payload<T>) => string) {
    webhook.on(event, async (callback) => {
      const { repository } = callback.payload
      const ids = config.repos[repository.full_name]
      if (!ids) return

      const message = handler(callback.payload)
      await ctx.broadcast(ids, message)
    })
  }

  function formatMarkdown(source: string) {
    return source.replace(/\n\s*\n/g, '\n')
  }

  registerHandler('commit_comment.created', ({ repository, comment }) => {
    return [
      `[GitHub] ${comment.user.login} commented on commit ${repository.full_name}@${comment.commit_id.slice(0, 6)}`,
      `URL: ${comment.html_url}`,
      formatMarkdown(comment.body),
    ].join('\n')
  })

  registerHandler('issues.opened', ({ repository, issue }) => {
    return [
      `[GitHub] ${issue.user.login} opened an issue ${repository.full_name}#${issue.number}`,
      `Title: ${issue.title}`,
      `URL: ${issue.html_url}`,
      formatMarkdown(issue.body),
    ].join('\n')
  })

  registerHandler('issue_comment.created', ({ comment, issue, repository }) => {
    const type = issue['pull_request'] ? 'pull request' : 'issue'
    return [
      `[GitHub] ${comment.user.login} commented on ${type} ${repository.full_name}#${issue.number}`,
      `URL: ${comment.html_url}`,
      formatMarkdown(comment.body),
    ].join('\n')
  })

  registerHandler('pull_request.opened', ({ repository, pull_request }) => {
    return [
      `[GitHub] ${pull_request.user.login} opened a pull request ${repository.full_name}#${pull_request.number} (${pull_request.base.label} <- ${pull_request.head.label})`,
      `URL: ${pull_request.html_url}`,
      formatMarkdown(pull_request.body),
    ].join('\n')
  })

  registerHandler('pull_request_review.submitted', ({ repository, review, pull_request }) => {
    if (!review.body) return
    return [
      `[GitHub] ${review.user.login} reviewed pull request ${repository.full_name}#${pull_request.number}`,
      `URL: ${review.html_url}`,
      // @ts-ignore
      formatMarkdown(review.body),
    ].join('\n')
  })

  registerHandler('pull_request_review_comment.created', ({ repository, comment, pull_request }) => {
    return [
      `[GitHub] ${comment.user.login} commented on pull request review ${repository.full_name}#${pull_request.number}`,
      `Path: ${comment.path}`,
      `URL: ${comment.html_url}`,
      formatMarkdown(comment.body),
    ].join('\n')
  })

  registerHandler('push', ({ compare, pusher, commits, repository, ref, after }) => {
    // do not show pull request merge
    if (/^0+$/.test(after)) return

    // use short form for tag releases
    if (ref.startsWith('refs/tags')) {
      return `[GitHub] ${pusher.name} published tag ${repository.full_name}@${ref.slice(10)}`
    }

    return [
      `[GitHub] ${pusher.name} pushed to ${repository.full_name}:${ref}`,
      `Compare: ${compare}`,
      ...commits.map(c => `[${c.id.slice(0, 6)}] ${formatMarkdown(c.message)}`),
    ].join('\n')
  })
}
