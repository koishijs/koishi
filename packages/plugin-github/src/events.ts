/* eslint-disable camelcase */

import { EventTypesPayload } from '@octokit/webhooks/dist-types/generated/get-webhook-payload-type-from-event'
import { EventData } from './server'

type WebhookEvent = Exclude<keyof EventTypesPayload, 'error'>

type SubEvent<T extends WebhookEvent> = {
  [K in WebhookEvent]: K extends '*' ? never : EventTypesPayload[K] extends EventTypesPayload[T] ? K : never
}[WebhookEvent]

interface CommentEventConfig {
  created?: boolean
  deleted?: boolean
  edited?: boolean
}

export interface EventConfig {
  commitComment?: boolean | CommentEventConfig
  create?: boolean
  delete?: boolean
  fork?: boolean
  issueComment?: boolean | CommentEventConfig
  issues?: boolean | {
    assigned?: boolean
    closed?: boolean
    deleted?: boolean
    demilestoned?: boolean
    edited?: boolean
    labeled?: boolean
    locked?: boolean
    milestoned?: boolean
    opened?: boolean
    pinned?: boolean
    reopened?: boolean
    transferred?: boolean
    unassigned?: boolean
    unlabeled?: boolean
    unlocked?: boolean
    unpinned?: boolean
  }
  pullRequest?: boolean | {
    assigned?: boolean
    closed?: boolean
    edited?: boolean
    labeled?: boolean
    locked?: boolean
    merged?: boolean
    opened?: boolean
    readyForReview?: boolean
    reopened?: boolean
    reviewRequestRemoved?: boolean
    reviewRequested?: boolean
    synchronize?: boolean
    unassigned?: boolean
    unlabeled?: boolean
    unlocked?: boolean
  }
  pullRequestReview?: boolean | {
    dismissed?: boolean
    edited?: boolean
    submitted?: boolean
  }
  pullRequestReviewComment?: boolean | CommentEventConfig
  push?: boolean
  star?: boolean | {
    created?: boolean
    deleted?: boolean
  }
}

export const defaultEvents: EventConfig = {
  commitComment: {
    created: true,
  },
  issueComment: {
    created: true,
  },
  issues: {
    closed: true,
    opened: true,
  },
  pullRequest: {
    closed: true,
    opened: true,
  },
  pullRequestReview: {
    submitted: true,
  },
  pullRequestReviewComment: {
    created: true,
  },
  star: {
    created: true,
  },
}

type Payload<T extends WebhookEvent> = EventTypesPayload[T]['payload']
type EventHandler<T extends WebhookEvent, P = {}> = (payload: Payload<T>) => EventData<P>

function formatMarkdown(source: string) {
  return source
    .replace(/^```(.*)$/gm, '')
    .replace(/^<!--(.*)-->$/gm, '')
    .replace(/\n\s*\n/g, '\n')
}

type FactoryCreator = <T extends WebhookEvent, P = {}>
  (callback: (event: T, payload: Payload<T>, handler: EventHandler<T, P>) => EventData<P>)
    => <E extends T>(event: E, handler?: EventHandler<E, P>) => void

export function addListeners(on: <T extends WebhookEvent>(event: T, handler: EventHandler<T>) => void) {
  const createFactory: FactoryCreator = callback => (event, handler) => on(event, payload => callback(event, payload, handler))

  type CommentEvent = 'commit_comment' | 'issue_comment' | 'pull_request_review_comment'

  interface CommentReplyPayloads {
    padding?: number[]
  }

  const onComment = createFactory<CommentEvent, CommentReplyPayloads>((event, payload, handler) => {
    const { user, body, html_url, url } = payload.comment
    if (user.type === 'Bot') return

    const [target, replies] = handler(payload)
    if (payload.action === 'deleted') {
      return [`${user.login} deleted a comment on ${target}`]
    }

    const { padding } = replies
    delete replies.padding

    const index = html_url.indexOf('#')
    const operation = payload.action === 'created' ? 'commented' : 'edited a comment'
    return [`${user.login} ${operation} on ${target}\n${formatMarkdown(body)}`, {
      link: [html_url],
      react: [url + `/reactions`],
      shot: [
        html_url.slice(0, index),
        html_url.slice(index),
        padding,
      ],
      ...replies,
    }]
  })

  onComment('commit_comment', ({ repository, comment }) => {
    const { full_name } = repository
    const { commit_id, path, position } = comment
    return [`commit ${full_name}@${commit_id.slice(0, 6)}\nPath: ${path}`, {
      // https://docs.github.com/en/rest/reference/repos#create-a-commit-comment
      reply: [`https://api.github.com/repos/${full_name}/commits/${commit_id}/comments`, { path, position }],
    }]
  })

  const onReference = createFactory<'create' | 'delete'>((event, { repository, ref, ref_type, sender }) => {
    const ref_name = `${repository.full_name}${ref_type === 'tag' ? '@' : ':'}${ref}`
    return [`${sender.login} ${event}d ${ref_type} ${ref_name}`]
  })

  onReference('create')

  onReference('delete')

  on('fork', ({ repository, sender, forkee }) => {
    const { full_name, forks_count } = repository
    return [`${sender.login} forked ${full_name} to ${forkee.full_name} (total ${forks_count} forks)`]
  })

  onComment('issue_comment', ({ issue, repository }) => {
    const { full_name } = repository
    const { number, comments_url } = issue
    const type = issue['pull_request'] ? 'pull request' : 'issue'
    return [`${type} ${full_name}#${number}`, {
      reply: [comments_url],
      padding: [16, 16, 16, 88],
    }]
  })

  const onIssue = createFactory<SubEvent<'issues'>>((event, payload, handler) => {
    const { user, url, html_url, comments_url } = payload.issue
    if (user.type === 'Bot') return

    const [message, replies] = handler(payload)
    return [message, {
      close: [url, comments_url],
      link: [html_url],
      react: [url + `/reactions`],
      reply: [comments_url],
      ...replies,
    }]
  })

  onIssue('issues.opened', ({ repository, issue, sender }) => {
    const { full_name } = repository
    const { title, body, number } = issue
    return [[
      `${sender.login} opened an issue ${full_name}#${number}`,
      `Title: ${title}`,
      formatMarkdown(body),
    ].join('\n')]
  })

  onIssue('issues.closed', ({ repository, issue, sender }) => {
    const { full_name } = repository
    const { title, number } = issue
    return [`${sender.login} closed issue ${full_name}#${number}\n${title}`]
  })

  onComment('pull_request_review_comment', ({ repository, comment, pull_request }) => {
    const { full_name } = repository
    const { number } = pull_request
    const { path, url } = comment
    return [`pull request review ${full_name}#${number}\nPath: ${path}`, {
      reply: [url],
    }]
  })

  on('pull_request_review.submitted', ({ repository, review, pull_request }) => {
    if (!review.body) return
    const { full_name } = repository
    const { number, comments_url } = pull_request
    const { user, html_url, body } = review
    if (user.type === 'Bot') return

    return [[
      `${user.login} reviewed pull request ${full_name}#${number}`,
      formatMarkdown(body),
    ].join('\n'), {
      link: [html_url],
      reply: [comments_url],
    }]
  })

  const onPullRequest = createFactory<SubEvent<'pull_request'>>((event, payload, handler) => {
    const { user, url, html_url, issue_url, comments_url } = payload.pull_request
    if (user.type === 'Bot') return

    const [message, replies] = handler(payload)
    return [message, {
      base: [url],
      close: [issue_url, comments_url],
      link: [html_url],
      merge: [url + '/merge'],
      rebase: [url + '/merge'],
      squash: [url + '/merge'],
      react: [issue_url + `/reactions`],
      reply: [comments_url],
      ...replies,
    }]
  })

  onPullRequest('pull_request.closed', ({ repository, pull_request, sender }) => {
    const { full_name } = repository
    const { title, number, merged } = pull_request
    const type = merged ? 'merged' : 'closed'
    return [`${sender.login} ${type} pull request ${full_name}#${number}\n${title}`]
  })

  onPullRequest('pull_request.opened', ({ repository, pull_request, sender }) => {
    const { full_name, owner } = repository
    const { title, base, head, body, number } = pull_request

    const prefix = new RegExp(`^${owner.login}:`)
    const baseLabel = base.label.replace(prefix, '')
    const headLabel = head.label.replace(prefix, '')
    return [[
      `${sender.login} opened a pull request ${full_name}#${number} (${baseLabel} ← ${headLabel})`,
      `Title: ${title}`,
      formatMarkdown(body),
    ].join('\n')]
  })

  on('push', ({ compare, pusher, commits, repository, ref, before, after }) => {
    const { full_name } = repository

    // do not show branch create / delete
    if (/^0+$/.test(before) || /^0+$/.test(after)) return

    return [[
      `${pusher.name} pushed to ${full_name}:${ref.replace(/^refs\/heads\//, '')}`,
      ...commits.map(c => `[${c.id.slice(0, 6)}] ${formatMarkdown(c.message)}`),
    ].join('\n'), {
      link: [compare],
    }]
  })

  on('star.created', ({ repository, sender }) => {
    const { full_name, stargazers_count } = repository
    return [`${sender.login} starred ${full_name} (total ${stargazers_count} stargazers)`]
  })
}
