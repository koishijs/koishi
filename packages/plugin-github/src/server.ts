/* eslint-disable camelcase */

import { Webhooks } from '@octokit/webhooks'
import { EventConfig } from './events'
import axios, { AxiosError, Method } from 'axios'
import { Session, User } from 'koishi-core'
import { CQCode, Logger } from 'koishi-utils'

declare module 'koishi-core/dist/database' {
  interface User {
    ghAccessToken?: string
    ghRefreshToken?: string
  }
}

User.extend(() => ({
  ghAccessToken: '',
  ghRefreshToken: '',
}))

export interface Config {
  secret?: string
  webhook?: string
  authorize?: string
  messagePrefix?: string
  appId?: string
  appSecret?: string
  redirect?: string
  promptTimeout?: number
  replyTimeout?: number
  requestTimeout?: number
  repos?: Record<string, number[]>
  events?: EventConfig
}

export interface OAuth {
  access_token: string
  expires_in: string
  refresh_token: string
  refresh_token_expires_in: string
  token_type: string
  scope: string
}

interface RequestOptions {
  url: string
  method: Method
  session: ReplySession
  body: any
  headers?: Record<string, any>
}

type ReplySession = Session<'ghAccessToken' | 'ghRefreshToken'>

const logger = new Logger('github')

export class GitHub extends Webhooks {
  constructor(public config: Config) {
    super({ ...config, path: config.webhook })
  }

  async getTokens(params: any) {
    const { data } = await axios.post<OAuth>('https://github.com/login/oauth/access_token', {
      client_id: this.config.appId,
      client_secret: this.config.appSecret,
      ...params,
    }, {
      headers: { Accept: 'application/json' },
    })
    return data
  }

  private async _request(url: string, method: Method, session: ReplySession, body: any, headers?: Record<string, any>) {
    logger.debug(method, url, body)
    await axios.post(url, body, {
      timeout: this.config.requestTimeout,
      headers: {
        accept: 'application/vnd.github.v3+json',
        authorization: `token ${session.$user.ghAccessToken}`,
        ...headers,
      },
    })
  }

  async authorize(session: Session, message: string) {
    await session.$send(message)
    const name = await session.$prompt(this.config.promptTimeout)
    if (!name) return session.$send('输入超时。')
    return session.$execute({ command: 'github.authorize', args: [name] })
  }

  async request(url: string, method: Method, session: ReplySession, body: any, headers?: Record<string, any>) {
    if (!session.$user.ghAccessToken) {
      return this.authorize(session, '要使用此功能，请对机器人进行授权。输入你的 GitHub 用户名。')
    }

    try {
      return await this._request(url, method, session, body, headers)
    } catch (error) {
      const { response } = error as AxiosError
      if (response?.status !== 401) {
        logger.warn(error)
        return session.$send('发送失败。')
      }
    }

    try {
      const data = await this.getTokens({
        refresh_token: session.$user.ghRefreshToken,
        grant_type: 'refresh_token',
      })
      session.$user.ghAccessToken = data.access_token
      session.$user.ghRefreshToken = data.refresh_token
    } catch {
      return this.authorize(session, '令牌已失效，需要重新授权。输入你的 GitHub 用户名。')
    }

    try {
      await this._request(url, method, session, body, headers)
    } catch (error) {
      logger.warn(error)
      return session.$send('发送失败。')
    }
  }
}

function formatReply(source: string) {
  return CQCode.parseAll(source).map((value) => {
    if (typeof value === 'string') return value
    if (value.type === 'image') return `![image](${value.data.url})`
    return ''
  }).join('')
}

type ReplyPayloads = {
  [K in keyof ReplyHandler]?: ReplyHandler[K] extends (...args: infer P) => any ? P : never
}

export type EventData<T = {}> = [string, (ReplyPayloads & T)?]

export class ReplyHandler {
  constructor(public github: GitHub, public session: Session, public content?: string) {}

  link(url: string) {
    return this.session.$send(url)
  }

  react(url: string) {
    return this.github.request(url, 'POST', this.session, {
      content: this.content,
    }, {
      accept: 'application/vnd.github.squirrel-girl-preview',
    })
  }

  reply(url: string, params?: Record<string, any>) {
    return this.github.request(url, 'POST', this.session, {
      body: formatReply(this.content),
      ...params,
    })
  }

  base(url: string) {
    return this.github.request(url, 'PATCH', this.session, {
      base: this.content,
    })
  }

  merge(url: string, method?: 'merge' | 'squash' | 'rebase') {
    const [title] = this.content.split('\n', 1)
    const message = this.content.slice(title.length)
    return this.github.request(url, 'PUT', this.session, {
      merge_method: method,
      commit_title: title.trim(),
      commit_message: message.trim(),
    })
  }

  rebase(url: string) {
    return this.merge(url, 'rebase')
  }

  squash(url: string) {
    return this.merge(url, 'squash')
  }

  async close(url: string, commentUrl: string) {
    if (this.content) await this.reply(commentUrl)
    await this.github.request(url, 'PATCH', this.session, {
      state: 'closed',
    })
  }

  async shot(url: string, selector: string, padding: number[] = []) {
    const page = await this.session.$app.browser.newPage()
    let buffer: Buffer
    try {
      await page.goto(url)
      const el = await page.$(selector)
      const clip = await el.boundingBox()
      const [top = 0, right = 0, bottom = 0, left = 0] = padding
      clip.x -= left
      clip.y -= top
      clip.width += left + right
      clip.height += top + bottom
      buffer = await page.screenshot({ clip })
    } catch (error) {
      new Logger('puppeteer').warn(error)
      return this.session.$send('截图失败。')
    } finally {
      await page.close()
    }
    return this.session.$send(`[CQ:image,file=base64://${buffer.toString('base64')}]`)
  }
}
