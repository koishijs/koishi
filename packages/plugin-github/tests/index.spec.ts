import { MockedApp } from 'koishi-test-utils'
import { apply, webhooks } from '../src'
import { readdirSync } from 'fs-extra'
import { resolve } from 'path'

const app = new MockedApp({
  githubWebhook: {
    secret: 'secret',
  },
})

app.plugin(apply, {
  'koishijs/koishi': [123],
})

const webhook = webhooks['/secret12140']

readdirSync(resolve(__dirname, '__fixtures__')).forEach((file) => {
  file = file.slice(0, -5)
  const [name] = file.split('.', 1)
  const payload = require(`./__fixtures__/${file}`)

  test(file, async () => {
    await webhook.receive({ id: 'id', name, payload })
    app.shouldMatchSnapshot(file)
  })
})
