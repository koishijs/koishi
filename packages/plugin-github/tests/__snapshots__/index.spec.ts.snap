// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check_run.completed: check_run.completed 1`] = `undefined`;

exports[`chect_suite.completed: chect_suite.completed 1`] = `undefined`;

exports[`commit_comment.created: commit_comment.created 1`] = `
Array [
  "send_group_msg",
  Object {
    "auto_escape": false,
    "group_id": 123,
    "message": "[GitHub] Commit Comment (koishijs/koishi)
User: Shigma
URL: https://github.com/koishijs/koishi/commit/bff469eabe14d42683a4f7c3ccb659daec5e1c00#commitcomment-36878220
This will introduce a failure in test.",
  },
]
`;

exports[`issue_comment.created.1: issue_comment.created.1 1`] = `
Array [
  "send_group_msg",
  Object {
    "auto_escape": false,
    "group_id": 123,
    "message": "[GitHub] Issue Comment (koishijs/koishi#19)
User: simon300000
URL: https://github.com/koishijs/koishi/issues/19#issuecomment-576277946
Mich würde auch interessieren, was ist „CoolQ“?",
  },
]
`;

exports[`issue_comment.created.2: issue_comment.created.2 1`] = `
Array [
  "send_group_msg",
  Object {
    "auto_escape": false,
    "group_id": 123,
    "message": "[GitHub] Pull Request Comment (koishijs/koishi#20)
User: Kouchya
URL: https://github.com/koishijs/koishi/pull/20#issuecomment-576291300
C'est important!",
  },
]
`;

exports[`issues.assigned: issues.assigned 1`] = `undefined`;

exports[`issues.closed: issues.closed 1`] = `undefined`;

exports[`issues.labeled: issues.labeled 1`] = `undefined`;

exports[`issues.opened: issues.opened 1`] = `
Array [
  "send_group_msg",
  Object {
    "auto_escape": false,
    "group_id": 123,
    "message": "[GitHub] Issue Opened (koishijs/koishi#19)
Title: Wie kann man um das Koishi zu installieren?
User: simon300000
URL: https://github.com/koishijs/koishi/issues/19
Ich verstecke Englisch und Chinesisch nicht! Gab es Personen, die mir helfen kann?",
  },
]
`;

exports[`pull_request.opened: pull_request.opened 1`] = `
Array [
  "send_group_msg",
  Object {
    "auto_escape": false,
    "group_id": 123,
    "message": "[GitHub] Pull Request Opened (koishijs/koishi#364852707)
koishijs:develop <- koishijs:experimental
User: simon300000
URL: https://github.com/koishijs/koishi/pull/20
Das ist wichtig!",
  },
]
`;

exports[`pull_request_review.submitted.1: pull_request_review.submitted.1 1`] = `undefined`;

exports[`pull_request_review.submitted.2: pull_request_review.submitted.2 1`] = `
Array [
  "send_group_msg",
  Object {
    "auto_escape": false,
    "group_id": 123,
    "message": "[GitHub] Pull Request Review (koishijs/koishi#364852707)
User: Kouchya
URL: https://github.com/koishijs/koishi/pull/20#pullrequestreview-345349537
LGTM",
  },
]
`;

exports[`pull_request_review_comment.created: pull_request_review_comment.created 1`] = `
Array [
  "send_group_msg",
  Object {
    "auto_escape": false,
    "group_id": 123,
    "message": "[GitHub] Pull Request Review Comment (koishijs/koishi#364852707)
Path: packages/test-utils/src/mocks.ts
User: Shigma
URL: https://github.com/koishijs/koishi/pull/20#discussion_r368570320
Naming is so hard......",
  },
]
`;

exports[`push.commit: push.commit 1`] = `
Array [
  "send_group_msg",
  Object {
    "auto_escape": false,
    "group_id": 123,
    "message": "[GitHub] Push (koishijs/koishi)
Ref: refs/heads/develop
User: Shigma
Compare: https://github.com/koishijs/koishi/compare/976c6e8f09a4...3ae7e7044d06
chore: adjust
fix(core): create major context at demand",
  },
]
`;

exports[`push.tag: push.tag 1`] = `
Array [
  "send_group_msg",
  Object {
    "auto_escape": false,
    "group_id": 123,
    "message": "[GitHub] koishijs/koishi published tag 1.5.0",
  },
]
`;
